File: day1/lua/a.lua
-- module for file operations
local open = io.open

-- taken from the INTERNET!
function string:split(delimiter)
	local result = {}
	local from = 1
	local delim_from, delim_to = string.find(self, delimiter, from, true)
	while delim_from do
		if delim_from ~= 1 then
			table.insert(result, string.sub(self, from, delim_from - 1))
		end
		from = delim_to + 1
		delim_from, delim_to = string.find(self, delimiter, from, true)
	end
	if from <= #self then
		table.insert(result, string.sub(self, from))
	end
	return result
end

local sum = 0

local file = open("input.txt", "rb")
if not file then
	return nil
end

-- M = {}
--
-- M = {
--     value = 10,
--     test = function(self)
--         self.value = self.value + 5
--         return self.value
--     end,
-- }
--
-- print(M:test()) // output = 15

-- : signifies call self within a table. `file` is a table. it is returned by the open function.
local content = file:read("*a")
file:close()

-- define the table that will hold each line
local lines = {}

-- operations seem to be done through calling the module of the data structure. ie: table, string, etc.
for line in content:gmatch("[^\r\n]*") do
	table.insert(lines, line)
end

local left_side = {}
local right_side = {}

for _, line in ipairs(lines) do
	local split_lines = line:split("   ")

	table.insert(left_side, split_lines[1])
	table.insert(right_side, split_lines[2])
end

-- sort both lists
table.sort(left_side, function(a, b)
	return a < b
end)
table.sort(right_side, function(a, b)
	return a < b
end)

-- for each value in the sides, subtract both, take the abs val then add to total
for index, value in ipairs(left_side) do
	local difference = math.abs(value - right_side[index])
	sum = sum + difference
end

print(sum)

File: day1/ts/a.ts
let sum: number = 0;

const input = await Deno.readTextFile("input.txt");

// console.log(input);

// define a left list and a right list
let left_list: number[] = [];
let right_list: number[] = [];

// split input by \n

const lines = input.split("\n");

// console.log(lines);

//  "65897   82748", "44296   15919", "21934   96178", "77571   39195",

// for every line
// split input by 3 spaces
for (const line of lines) {
    const trim = line.trim();
    const lists_content = trim.split("   ");

    // put first index in left list, second index in right list
    if (lists_content.length === 2) {
        const leftValue = Number(lists_content[0]);
        const rightValue = Number(lists_content[1]);

        // without the isNaN checks, i would get an undefined somewhere ?
        if (!isNaN(leftValue) && !isNaN(rightValue)) {
            left_list.push(Number(lists_content[0]));
            right_list.push(Number(lists_content[1]));
        }
    }
}

// after all inputs are in, sort the two lists by least to greatest
left_list.sort((a, b) => a - b);
right_list.sort((a, b) => a - b);

// for the list size, find the difference of the index by math.abs and then add that to the sum
for (let i = 0; i < left_list.length; i++) {
    sum += Math.abs(left_list[i] - right_list[i]);
    console.log(
        `Left value: ${left_list[i]}, right value: ${right_list[i]}\nEqual to:`,
        sum,
    );
}

console.log(sum);
